### DISCUSSION TITLE
What is the best way to simulate teleportation of a dynamic object in dyn4j?

### DISCUSSION THREAD

**Discussion by LordXerus** [dyn4j/dyn4j#224] [Q&A] [answered]
Essentially directly manipulating the position and velocity of a dynamic body. Do I need to delete and recreate the body?

**Comment by wnbittle** [comment, marked as answer]
Do you mean controlling a body by changing it's position and velocity or true on demand teleportation?  Either way, both are possible by just using the setters `Body.setVelocity(...)` or `Body.translate(...)`.

This is definitely a tricky thing to do.  I would go about it this way (again, assuming you are not doing this to control a body over time, but instead treating as a true teleport from one place to another):

```java
Body b = //
b.translate(...);
// you want to reset the previous transform if you are using CCD
b.getPreviousTransform().set(b.getTransform());
```

**Reply by LordXerus** [reply]
Yep, this is an instantaneous teleport. If I want to control a Body over time, I would have to set the MassType to Infinite so it's excluded from the physics, right?

**Comment by wnbittle** [comment]
There are other ways to control the body without direct manipulation (see the samples repo), but this is one way yes.  The challenge with this approach is that it'll steam roll everything because it's infinite mass.  The body does not need to be infinite for you to teleport it, if you do want it to participate in physics still.

**Reply by LordXerus** [reply]
Ok so right now I know of 3 useful ways to manipulate the Body.

1. `translate` + `getPreviousTransform().set`. Used for instantaneous teleportation. Repeated use will cause undesirable behavior? Also what's the difference between `getTransform.set()` and `translate()`. Is it only absolute vs relative?
2. `setLinearVelocity(deltaPos / deltaTime)`. I assume this makes it easier for the simulation to run. This combined with Infinite mass is a good method to use if the Body should follow the mouse exactly, right? Sometimes I don't care about participating in physics as much as making a Body that allows a user to interact with the physics. What happens if an infinite mass object collides with a dynamic object? From my understanding of the conservation of momentum it would send the dynamic object away at infinite speed... which isn't good.
3. `PinJoint`. Use this unary joint when dealing with dynamic objects right? Do I need to increase the frequency to make it follow the mouse more?

Are my descriptions and assumptions correct? Are there any other common ways of manipulating a Body's transform? What's the preferred method

**Comment by wnbittle** [comment]
1. Right.  Using `getPreviousTransform().set` makes sure the previous and current transforms are identical.  Translate applies an x/y movement from the current position.
2. If you want an immovable object, but it still interact, then setting the linear velocity would work.  If it runs into a dynamic body it would resolve the collision by sending the dynamic body away.  The speed will be dependent on two factors, the dynamic body and immovable body's _relative_ velocity and the coefficient of restitution.  It should not send it off at an infinite speed.
3. Take a look at the [samples project](https://github.com/dyn4j/dyn4j-samples).  One of them uses the `MotorJoint` to control a body, but still have it interact with gravity and other objects normally.

The preferred method depends on what you are trying to do really.  Not knowing what you want to achieve, the `MotorJoint` would be my recommendation.
